util-images/probes/pkg/dnspropagation/probe.go

```diff
@@ -0,0 +1,242 @@
+
+	if len(durations) == 0 {
+		klog.Warningf("DNS propagation probe has zero obseravtions")
+		return
+	}
+	sum := 0.0
+	for _, duration := range durations {
+		sum += duration
+	}
+	klog.V(1).Infof("DNS propagation probe finished, total of %v observations, average duration, %v s", len(durations), sum/float64(len(durations)))
+}
+
+func selectSample(podCount int, sampleCount int) []int {
+	indices := make([]int, podCount)
+	for idx := range indices {
+		indices[idx] = idx
+	}
+	rand.Shuffle(len(indices), func(i, j int) { indices[i], indices[j] = indices[j], indices[i] })
+	indices = indices[:sampleCount]
+	return indices
+}
+func runSinglePod(client kubernetes.Interface, url string, podName string, namespace string, interval time.Duration) time.Duration {
+	klog.V(4).Infof("Starting dns propagation calculation for pod %s ...", url)
+	tick := time.NewTicker(interval)
+	for {
+		select {
+		case <-tick.C:
+			klog.V(4).Infof("DNS lookup %s", url)
+			if err := lookup(url); err != nil {
+				klog.Warningf("DNS lookup error: %v", err)
+				continue
+			}
+			endTime := time.Now()
+			klog.V(2).Infof("DNS lookup finished for pod %s, finding pod running time...", url)
+			timestamp, err := fetchPodRunningTime(client, podName, namespace)
+			if err != nil {
+				klog.Fatalf("K8s error: %v", err)
+				continue
+			}
+			duration := endTime.Sub(timestamp)
+			klog.V(2).Infof("Pod running time fetched for pod %s, timestamp= %v, DNS propagation duration= %v s", url, timestamp, duration)
+			return duration
+		}
+	}
+}
+
+func lookup(url string) error {
+	_, err := net.LookupIP(url)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+func fetchPodRunningTime(client kubernetes.Interface, podName string, namespace string) (time.Time, error) {
+	trials := 0
+	for trials < 3 {
+		readyTimestamp, err := getPodRunningTimeFromClient(client, podName, namespace, true)
+		if err == nil {
+			return readyTimestamp, nil
+		}
+		klog.Warningf("Failed at obtaining pod running time for pod %s, reason: %v", podName, err)
+		trials++
+		time.Sleep(1 * time.Second)
+	}
+	return getPodRunningTimeFromClient(client, podName, namespace, false)
+}
+
+func getNamespacesFromClient(client kubernetes.Interface, namespace string) (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+	namespaceList, err := client.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return "", err
+	}
+	for _, namespaceItem := range namespaceList.Items {
+		if strings.HasPrefix(namespaceItem.Name, namespace) {
+			return namespaceItem.Name, nil
+		}
+	}
+	return "", errors.New("No namespace were found that matches the prefix: " + namespace)
+}
+
+func getPodRunningTimeFromClient(client kubernetes.Interface, podName string, namespace string, resourceVersionEnabled bool) (time.Time, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
```

Comment:
this is too short, also the client has retries and timeouts by default, any particular reason you want to cancel the context?

Reply:
I'm not so sure - single object GETs (especially from the watch cache) should be pretty quick.

Our SLO threshold for API call latency for single object API calls is 1s in 99th percentile, so I think 3s is a pretty sane assumption here as some additional buffer.

If we go above that value, then the cluster is definitely overloaded and not really usable anyway :P.

Reply:
Also, why not use a watch here instead of periodic polling?